raft/src/main/java/org/apache/kafka/raft/LeaderState.java

```diff
@@ -202,7 +214,46 @@ public void appendLeaderChangeMessage(long currentTimeMs) {
             .setVoters(voters)
             .setGrantingVoters(grantingVoters);
 
-        accumulator.appendLeaderChangeMessage(leaderChangeMessage, currentTimeMs);
+        accumulator.appendControlMessages((baseOffset, epoch, buffer) -> {
+            try (MemoryRecordsBuilder builder = new MemoryRecordsBuilder(
+                buffer,
+                RecordBatch.CURRENT_MAGIC_VALUE,
+                Compression.NONE,
+                TimestampType.CREATE_TIME,
+                baseOffset,
+                currentTimeMs,
+                RecordBatch.NO_PRODUCER_ID,
+                RecordBatch.NO_PRODUCER_EPOCH,
+                RecordBatch.NO_SEQUENCE,
+                false,
+                true,
+                epoch,
+                buffer.capacity()
+            )
+            ) {
+                builder.appendLeaderChangeMessage(
+                    currentTimeMs,
+                    leaderChangeMessage
+                );
+                VoterSetOffset voterSetOffset = lastVoterSetOffset.orElse(null);
+                // if lastVoterOffset is -1 we know the leader hasn't written the bootstrap snapshot records to the log yet
+                if (voterSetOffset != null && voterSetOffset.offset() == -1) {
+                    if (kraftVersion > 0) {
```

Comment:
It is an `IllegalStateException` if a voters set exist in the bootstrap checkpoint but the kraft.version is 0.